### PR TITLE
Catch some incorrect method parameter or call argument formatting patterns with checkstyle

### PR DESCRIPTION
**Pull request by leventov**
There is no existing and precise check for this in checkstyle, but it's possible to create a regexp that catches a lot of violations of the corresponding Druid style rule without false positives.

### CODE REVIEW

**Comment by b-slim** [comment]
+1

File: server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java [hunk 1]
```diff
@@ -170,17 +170,21 @@ public void configure(Binder binder)
         CURATOR_KEY
     );
 
+    // For some reason the second argument of Key.get() should be Annotation rather than Class, otherwise fails with NPE
```
Comment (b-slim) [inline server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java]: @leventov this is strange, what is the change that is causing this?

Comment (leventov) [reply server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java]: Replacing `() -> Coordinator.class` (a lambda of `Annotation` type) with just `Coordinator.class` in a `Key.get()` call (which has both overloadings) produced

```
1) Error in custom provider, java.lang.NullPointerException: annotation
  at org.apache.druid.guice.PolyBind.createChoiceWithDefault(PolyBind.java:109) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> org.apache.druid.curator.discovery.DiscoveryModule)
  while locating org.apache.druid.discovery.DruidLeaderSelector annotated with @org.apache.druid.client.coordinator.Coordinator()
    for the 17th parameter of org.apache.druid.server.coordinator.DruidCoordinator.<init>(DruidCoordinator.java:152)
  at org.apache.druid.cli.CliCoordinator$1.configure(CliCoordinator.java:177) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> org.apache.druid.cli.CliCoordinator$1)
  while locating org.apache.druid.server.coordinator.DruidCoordinator
1 error
	at com.google.inject.internal.InjectorImpl$2.get(InjectorImpl.java:1028) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1050) ~[guice-4.1.0.jar:?]
	at org.apache.druid.guice.LifecycleModule$2.start(LifecycleModule.java:132) ~[druid-core-0.13.0-incubating-SNAPSHOT.jar:0.13.0-incubating-SNAPSHOT]
	at org.apache.druid.cli.GuiceRunnable.initLifecycle(GuiceRunnable.java:105) [druid-services-0.13.0-incubating-SNAPSHOT.jar:0.13.0-incubating-SNAPSHOT]
	at org.apache.druid.cli.ServerRunnable.run(ServerRunnable.java:51) [druid-services-0.13.0-incubating-SNAPSHOT.jar:0.13.0-incubating-SNAPSHOT]
	at org.apache.druid.cli.Main.main(Main.java:118) [druid-services-0.13.0-incubating-SNAPSHOT.jar:0.13.0-incubating-SNAPSHOT]
Caused by: java.lang.NullPointerException: annotation
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:229) ~[guava-16.0.1.jar:?]
	at com.google.inject.Key.strategyFor(Key.java:356) ~[guice-4.1.0.jar:?]
	at com.google.inject.Key.get(Key.java:270) ~[guice-4.1.0.jar:?]
	at org.apache.druid.guice.PolyBind$ConfiggedProvider.get(PolyBind.java:179) ~[druid-core-0.13.0-incubating-SNAPSHOT.jar:0.13.0-incubating-SNAPSHOT]
	at com.google.inject.internal.ProviderInternalFactory.provision(ProviderInternalFactory.java:81) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalFactoryToInitializableAdapter.provision(InternalFactoryToInitializableAdapter.java:53) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderInternalFactory.circularGet(ProviderInternalFactory.java:61) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalFactoryToInitializableAdapter.get(InternalFactoryToInitializableAdapter.java:45) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:38) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:62) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:110) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:90) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:268) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1092) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40) ~[guice-4.1.0.jar:?]
	at org.apache.druid.guice.LifecycleScope$1.get(LifecycleScope.java:68) ~[druid-core-0.13.0-incubating-SNAPSHOT.jar:0.13.0-incubating-SNAPSHOT]
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl$2$1.call(InjectorImpl.java:1019) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1085) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl$2.get(InjectorImpl.java:1015) ~[guice-4.1.0.jar:?]
	... 5 more
```

But I have no idea why

**Comment by leventov** [comment]
@QiuMM No

File: server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java (hunk 1 above)
Comment (leventov) [reply server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java]: @b-slim does this comment cancel your previous +1 or not?

Comment (b-slim) [reply server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java]: not it does not, but would be nice if we have an answer why it is not working after this patch.

Comment (leventov) [reply server/src/main/java/org/apache/druid/curator/discovery/DiscoveryModule.java]: Ok, I think this was due to a bug in `PolyBind`. Should be fixed in https://github.com/apache/incubator-druid/pull/6461/commits/f9223557e96f25410a061b73f24058038ab66a85

**Comment by jihoonson** [comment]
Hmm, is Travis running for the last two commits? The last Travis job I found is https://travis-ci.org/apache/incubator-druid/builds/442767166 which tested https://github.com/apache/incubator-druid/commit/4bd3e4e29fd9ad319fc4c305e796ade53cdd8cad.

**Comment by leventov** [comment]
@jihoonson there is a Github outage: https://status.github.com/messages

**Comment by jihoonson** [comment]
@leventov thanks for the check. It's running now.

**Pull request merged**